File: src/stores/collection.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'

export interface OwnedTrainee {
  id: string
  nickname: string | null
  star_rank: number
  potential_spd: number
  potential_sta: number
  potential_pow: number
  potential_gut: number
  potential_wit: number
  inherited_skills: string
  parent_slot_1: string | null
  parent_slot_2: string | null
  notes: string | null
}

export interface OwnedCard {
  id: string
  card_level: number
  limit_break: number
  bond_level: number
  notes: string | null
}

export const useCollectionStore = defineStore('collection', () => {
  const trainees = ref<OwnedTrainee[]>([])
  const cards = ref<OwnedCard[]>([])
  const loading = ref(false)

  async function fetchCollection() {
    loading.value = true
    try {
      const data = await invoke<{ trainees: OwnedTrainee[], cards: OwnedCard[] }>('get_collection')
      trainees.value = data.trainees
      cards.value = data.cards
    } catch (e) {
      console.error('Failed to fetch collection', e)
    } finally {
      loading.value = false
    }
  }

  async function upsertTrainee(trainee: OwnedTrainee) {
    await invoke('upsert_trainee', { data: trainee })
    await fetchCollection()
  }

  async function upsertCard(card: OwnedCard) {
    await invoke('upsert_card', { data: card })
    await fetchCollection()
  }

  return {
    trainees,
    cards,
    loading,
    fetchCollection,
    upsertTrainee,
    upsertCard
  }
})

File: src-tauri/src/commands/collection.rs
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct OwnedTrainee {
    pub id: String,
    pub nickname: Option<String>,
    pub star_rank: u32,
    pub potential_spd: u32,
    pub potential_sta: u32,
    pub potential_pow: u32,
    pub potential_gut: u32,
    pub potential_wit: u32,
    pub inherited_skills: String, // JSON
    pub parent_slot_1: Option<String>,
    pub parent_slot_2: Option<String>,
    pub notes: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct OwnedCard {
    pub id: String,
    pub card_level: u32,
    pub limit_break: u32,
    pub bond_level: u32,
    pub notes: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Collection {
    pub trainees: Vec<OwnedTrainee>,
    pub cards: Vec<OwnedCard>,
}

#[tauri::command]
pub async fn get_collection() -> Result<Collection, String> {
    // This will be implemented using tauri-plugin-sql in a real app,
    // but for now, we'll return an empty collection as a placeholder.
    Ok(Collection {
        trainees: vec![],
        cards: vec![],
    })
}

#[tauri::command]
pub async fn clear_collection() -> Result<(), String> {
    println!("Clearing whole collection...");
    Ok(())
}

#[tauri::command]
pub async fn upsert_trainee(data: OwnedTrainee) -> Result<(), String> {
    println!("Upserting trainee: {:?}", data);
    Ok(())
}

#[tauri::command]
pub async fn upsert_card(data: OwnedCard) -> Result<(), String> {
    println!("Upserting card: {:?}", data);
    Ok(())
}
